docs/getting-started/walkthrough.md

```diff
@@ -24,177 +24,411 @@ specific language governing permissions and limitations
+One popular external system is [Apache Kafka](http://kafka.apache.org/).
+As a first step, lets update our function in `greeter/greeter.py` to respond to each input by sending a greeting to a Kafka topic.
 
-{% highlight java %}
-package org.apache.flink.statefun.examples.greeter;
+{% highlight python %}
+from messages_pb2 import GreetRequest, GreetResponse
+from statefun import StatefulFunctions
 
-import org.apache.flink.statefun.examples.kafka.generated.GreetRequest;
-import org.apache.flink.statefun.sdk.io.Router;
+functions = StatefulFunctions()
 
-final class GreetRouter implements Router<GreetRequest> {
+@functions.bind("example/greeter")
+def greet(context, message: GreetRequest):
+    response = GreetResponse()
+    response.name = message.name
+    response.greeting = "Hello {}".format(message.name)
+    
+    egress_message = kafka_egress_record(topic="greetings", key=message.name, value=response)
+    context.pack_and_send_egress("example/greets", egress_message)
+{% endhighlight %} 
 
-    @Override
-    public void route(GreetRequest message, Downstream<GreetRequest> downstream) {
-        downstream.forward(GreetingConstants.GREETER_FUNCTION_TYPE, message.getWho(), message);
-    }
-}
+For each message, a response is constructed and sent to a kafka topic call `greetings` partitioned by `name`.
+The `egress_message` is sent to a an `egress` named `example/greets`.
+This identifier points to a particular Kafka cluster and is configured on deployment below.
+
+## A Stateful Hello
+
+This is a great start, but does not show off the real power of stateful functions - working with state.
+Suppose you want to generate a personalized response for each user depending on how many times they have sent a request.
+
+{% highlight python %}
+def compute_greeting(name, seen):
+    """
+    Compute a personalized greeting, based on the number of times this @name had been seen before.
+    """
+    templates = ["", "Welcome %s", "Nice to see you again %s", "Third time is a charm %s"]
+    if seen < len(templates):
+        greeting = templates[seen] % name
+    else:
+        greeting = "Nice to see you at the %d-nth time %s!" % (seen, name)
+
+    response = GreetResponse()
+    response.name = name
+    response.greeting = greeting
+
+    return response
 {% endhighlight %}
 
-So, if a message for a user named John comes in, it will be shipped to John’s dedicated Greeter function.
-In case there is a following message for a user named Jane, a new instance of the Greeter function will be spawned.
+To “remember” information across multiple greeting messages, you then need to associate a persisted value field (``seen_count``) to the Greet function.
+For each user, functions can now track how many times they have been seen.
 
-## Persistence
+{% highlight python %}
+@functions.bind("example/greeter")
+def greet(context, message: GreetRequest):
```

Comment:
```suggestion
def greet(context, greet_request: GreetRequest):
```

Because I get 
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 2446, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1951, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1820, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1949, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.7/site-packages/flask/app.py", line 1935, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/app/greeter.py", line 72, in handle
    response_data = handler(request.data)
  File "/usr/local/lib/python3.7/site-packages/statefun/request_reply.py", line 39, in __call__
    reply = self.handle_invocation(request)
  File "/usr/local/lib/python3.7/site-packages/statefun/request_reply.py", line 54, in handle_invocation
    self.invoke_batch(batch, context, target_function)
  File "/usr/local/lib/python3.7/site-packages/statefun/request_reply.py", line 99, in invoke_batch
    fun(context, invocation.argument)
  File "/app/greeter.py", line 52, in greet
    response = compute_greeting(greet_request.name, state.seen)
NameError: name 'greet_request' is not defined
```

Reply:
nice catch